File: app/src/main/java/com/lucasgoudin/pictura/Filter/Convolution.java
package com.lucasgoudin.pictura.Filter;

import android.graphics.Bitmap;
import android.graphics.Color;

public class Convolution {

    static final int rgbMax = 255;
    static void ApplyConvolution(Bitmap bmp, int[] filter, int sizeFilter){
        //recover the information about the picture + pixel
        int w = bmp.getWidth();
        int h = bmp.getHeight();
        int [] pixels = new int [w * h];
        bmp.getPixels(pixels, 0, w, 0,0,w,h);

        //kernel
        int halfSize = (sizeFilter-1)/2;
        int indexF; //index of the neighbor

        for(int y=0; y<h; y++){
            for(int x=0; x<w; x++){
                int index = y * w + x;
                int cptF=0; //index for the kernel

                //recover information about the pixel x,y
                int pixel = pixels[index];
                int r = Color.red(pixel);
                int g = Color.green(pixel);
                int b = Color.blue(pixel);

                for(int i=-halfSize; i<=halfSize; i++){ //y = lines
                    for(int j=-halfSize; j<=halfSize; j++){ //x = columns
                        indexF = (y+i) * w + x+j;
                        if(indexF>=0 && indexF<w*h){    //check if index valid (in the picture)
                            //recover information about the pixel x+j,y+i
                            int pixelF = pixels[indexF];
                            int rF = Color.red(pixelF);
                            int gF = Color.green(pixelF);
                            int bF = Color.blue(pixelF);

                            r += rF*filter[cptF];
                            g += gF*filter[cptF];
                            b += bF*filter[cptF];

                        }
                        cptF++;
                    }
                }
                r = Math.min(255, Math.max(0,r));
                g = Math.min(255, Math.max(0,g));
                b = Math.min(255, Math.max(0,b));

                pixels [index] = Color.rgb(r,g,b);
            }
        }
        bmp.setPixels(pixels,0,w,0,0,w,h);
    }
}

File: app/src/main/rs/Convolution.rs
#pragma  version (1)
#pragma rs java_package_name(com.lucasgoudin.pictura)
#pragma rs_fp_relaxed

int w;
int h;
int *pixels;
int size;
int* filter;

uchar4 __attribute__((kernel)) convolution(uchar4 in,uint32_t x,uint32_t y) {
    int cptF=0; //index of the filter
    int indexF; //index of the neighbor

    //recover the Kernel's coefficient
    int sumF=0;
    for(int i=0; i < size*size; i++){
        sumF += filter[i];
    }
    //uchar4 same= rsGetElementAt_uchar4(in, x,y);
    float4 color=0;
    int l =(size-1)/2;
    for (int i=-l;i<=l;i++){   //row - y
        for(int j=-l;j<=l;j++){    //col - x
            indexF=(y+i)*w + x + j;
            if(indexF >= 0 && indexF < w*h){    //check if index valid (in the picture)
                float4 temp = rsUnpackColor8888(pixels[indexF]);
                color.r +=  (filter[cptF] * temp.r)/sumF;
                color.g +=  (filter[cptF] * temp.g)/sumF;
                color.g +=  (filter[cptF] * temp.b)/sumF;
                cptF++;
            }
        }
    }
    if(color.r > 255){
        color.r = 255;
    }
    if(color.g > 255){
        color.g = 255;
    }
    if(color.b > 255){
        color.b = 255;
    }
    return rsPackColorTo8888(color);
}


